coderd/healthcheck/provisioner.go

```diff
@@ -0,0 +1,141 @@
+	r.Dismissed = opts.Dismissed
+	now := opts.TimeNowFn()
+	if opts.StaleInterval == 0 {
+		opts.StaleInterval = provisionerdserver.DefaultHeartbeatInterval * 3
+	}
+
+	if opts.CurrentVersion == "" {
+		r.Severity = health.SeverityError
+		r.Error = ptr.Ref("Developer error: CurrentVersion is empty!")
+		return
+	}
+
+	if opts.CurrentAPIVersion == nil {
+		r.Severity = health.SeverityError
+		r.Error = ptr.Ref("Developer error: CurrentAPIVersion is nil!")
+		return
+	}
+
+	if opts.ProvisionerDaemonsFn == nil {
+		r.Severity = health.SeverityError
+		r.Error = ptr.Ref("Developer error: ProvisionerDaemonsFn is nil!")
+		return
+	}
+
+	// nolint: gocritic // need an actor to fetch provisioner daemons
+	daemons, err := opts.ProvisionerDaemonsFn(dbauthz.AsSystemRestricted(ctx))
+	if err != nil {
+		r.Severity = health.SeverityError
+		r.Error = ptr.Ref("error fetching provisioner daemons: " + err.Error())
+		return
+	}
+
+	for _, daemon := range daemons {
+		r.ProvisionerDaemons = append(r.ProvisionerDaemons, convertProvisionerDaemon(daemon))
+	}
+
+	if len(r.ProvisionerDaemons) == 0 {
+		r.Severity = health.SeverityError
+		r.Error = ptr.Ref("No provisioner daemons found!")
+		return
+	}
+
+	for _, daemon := range daemons {
+		// Daemon never connected, skip.
+		if !daemon.LastSeenAt.Valid {
+			continue
+		}
+		// Daemon has gone away, skip.
+		if now.Sub(daemon.LastSeenAt.Time) > (opts.StaleInterval) {
+			continue
+		}
+		// For release versions, just check MAJOR.MINOR and ignore patch.
+		if !semver.IsValid(daemon.Version) {
+			if r.Severity.Value() < health.SeverityWarning.Value() {
+				r.Severity = health.SeverityWarning
+			}
+			r.Warnings = append(r.Warnings, health.Messagef(health.CodeUnknown, "Provisioner daemon %q reports invalid version %q", opts.CurrentVersion, daemon.Version))
+		} else if !buildinfo.VersionsMatch(opts.CurrentVersion, daemon.Version) {
+			if r.Severity.Value() < health.SeverityWarning.Value() {
+				r.Severity = health.SeverityWarning
+			}
+			r.Warnings = append(r.Warnings, health.Messagef(health.CodeProvisionerDaemonVersionMismatch, "Provisioner daemon %q has outdated version %q", daemon.Name, daemon.Version))
+		}
+
+		// Provisioner daemon API version follows different rules.
+		if _, _, err := apiversion.Parse(daemon.APIVersion); err != nil {
+			if r.Severity.Value() < health.SeverityError.Value() {
+				r.Severity = health.SeverityError
+			}
+			r.Warnings = append(r.Warnings, health.Messagef(health.CodeUnknown, "Provisioner daemon %q reports invalid API version: %s", daemon.Name, err.Error()))
+		} else if err := opts.CurrentAPIVersion.Validate(daemon.APIVersion); err != nil {
+			if r.Severity.Value() < health.SeverityError.Value() {
+				r.Severity = health.SeverityError
+			}
+			r.Warnings = append(r.Warnings, health.Messagef(health.CodeProvisionerDaemonAPIVersionIncompatible, "Provisioner daemon %q reports incompatible API version: %s", daemon.Name, err.Error()))
```

Comment:
This seems unlikely; it doesn't seem right to insert it into the ProvisionerDaemons table if it connects with an incompatible version, since we aren't able to serve that connection and should close it.

I was thinking more along the lines of warning if the daemon isn't on the current major version, since that's in danger of being deprecated by an upgrade.